src/cmd/decode.rs

```diff
@@ -0,0 +1,163 @@
+pub struct DecodeCommand {
+    /// Type of data
+    #[clap(arg_enum, short, long)]
+    r#type: DataType,
+    /// The data to decode
+    #[clap(long)]
+    data: String,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, clap::ArgEnum)]
+enum DataType {
+    Event,
+    Message,
+}
+
+impl DecodeCommand {
+    pub fn run(&self) -> Result<()> {
+        let (_, contract_metadata) = load_metadata(None)?;
+        let transcoder = ContractMessageTranscoder::new(&contract_metadata);
+
+        const ERR_MSG: &str = "Failed to decode specified data as a hex value";
+        let decoded_data = match self.r#type {
+            DataType::Event => {
+                transcoder.decode_contract_event(&mut &decode_hex(&self.data).context(ERR_MSG)?[..])
+            }
+            DataType::Message => transcoder
+                .decode_contract_message(&mut &decode_hex(&self.data).context(ERR_MSG)?[..]),
+        };
+
+        println!(
+            "{:>width$} {:?}",
+            "Decoded data:".bright_green().bold(),
+            decoded_data,
+            width = DEFAULT_KEY_COL_WIDTH
+        );
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::util::tests::with_new_contract_project;
+    use std::path::Path;
+
+    /// Create a `cargo contract` command
+    fn cargo_contract(path: &Path) -> assert_cmd::Command {
+        let mut cmd = assert_cmd::Command::cargo_bin("cargo-contract").unwrap();
+        cmd.current_dir(path).arg("contract");
+        cmd
+    }
+
+    #[test]
+    fn decode_works() {
+        // given
+        let contract = r#"
+	#![cfg_attr(not(feature = "std"), no_std)]
+
+	use ink_lang as ink;
+
+	#[ink::contract]
+	pub mod switcher {
+	    #[ink(event)]
+	    pub struct Switched {
+		new_value: bool,
+	    }
+
+	    #[ink(storage)]
+	    pub struct Switcher {
+		value: bool,
+	    }
+
+	    impl Switcher {
+		#[ink(constructor)]
+		pub fn new(init_value: bool) -> Self {
+		    Self { value: init_value }
+		}
+
+		#[ink(message, selector = 0xBABEBABE)]
+		pub fn switch(&mut self, value: bool) {
+		    self.value = value;
+		}
+	    }
+	}"#;
+
+        // when
+        // contract is built
+        with_new_contract_project(|manifest_path| {
+            let project_dir = manifest_path.directory().expect("directory must exist");
+            let lib = project_dir.join("lib.rs");
+            std::fs::write(&lib, contract)?;
+
+            assert_cmd::Command::new("rustup")
+                .arg("override")
+                .arg("set")
+                .arg("nightly")
+                .assert()
+                .success();
+
+            log::info!("Building contract in {}", project_dir.to_string_lossy());
+            cargo_contract(project_dir).arg("build").assert().success();
+
+            let msg_data: &str = "babebabe01";
+            let msg_decoded: &str =
+                r#"Ok(Map(Map { ident: Some("switch"), map: {String("value"): Bool(true)} }))"#;
```

Comment:
Ok so your test has highlighted an issue here :tada:  We should not be printing the debug `{:?}` output but rather the `Display` output